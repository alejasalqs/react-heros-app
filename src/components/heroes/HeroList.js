import React from "react";
import { getHeroByPublisher } from "../../selectos/getHeroesByPublisher";

// por props enviamos el publisher con el cual filtramos dc o marvel
export const HeroList = ({ publisher }) => {
  const heros = getHeroByPublisher(publisher);
  return (
    <ul>
      {heros.map((hero) => (
        <li key={hero.id}>{hero.superhero}</li>
      ))}
    </ul>
  );
};
