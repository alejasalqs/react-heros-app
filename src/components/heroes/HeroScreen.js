import React, { useMemo } from "react";
import { Redirect, useParams } from "react-router";
import { getHeroById } from "../../selectos/getHeroByID";

export const HeroScreen = ({ history }) => {
  // Hook de react.router que nos permite obtener los paramentros del url
  // retorna un objeto por lo que se puede desestructurar
  const { heroId } = useParams();

  //const hero = getHeroById(heroId);
  const hero = useMemo(() => getHeroById(heroId), [heroId]);

  if (!hero) {
    // Como buena practica un componente siempre deve devolver algo que no sea un error
    // En este caso si el heroes no existe nos devuleve un redirect a otra pagina
    return <Redirect to="/marvel" />;
  }

  const {
    superhero,
    alter_ego,
    first_appearance,
    characters,
    publisher,
  } = hero;

  const handleReturn = () => {
    history.goBack();
  };

  return (
    <div className="row mt-5">
      <div className="col-4">
        <img
          src={`../Assets/heroes/${heroId}.jpg`}
          className="img-thumbnail"
          alt={superhero}
        />
      </div>
      <div className="col-8">
        <h3>{superhero}</h3>
        <ul className="list-group list-group-flush">
          <li className="list-group-item">
            <b>Alter ego: </b>
            {alter_ego}
          </li>
          <li className="list-group-item">
            <b>Publisher: </b>
            {publisher}
          </li>
          <li className="list-group-item">
            <b>Firts apperance: </b>
            {first_appearance}
          </li>
        </ul>
        <h5>Characters:</h5>
        <p>{characters}</p>
        <button className="btn btn-outline-info" onClick={handleReturn}>
          Return
        </button>
      </div>
    </div>
  );
};
