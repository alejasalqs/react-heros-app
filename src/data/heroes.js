export const heroes = [
  {
    id: "dc-batman",
    superhero: "Batman",
    publisher: "DC Comics",
    alter_ego: "Bruce Wayne",
    first_appearance: "Detective Comics #27",
    characters: "Bruce Wayne",
    img:
      "https://res.cloudinary.com/asalguerodev/image/upload/v1620275828/heros-app/dc-batman_fkzxn9.jpg",
  },
  {
    id: "dc-superman",
    superhero: "Superman",
    publisher: "DC Comics",
    alter_ego: "Kal-El",
    first_appearance: "Action Comics #1",
    characters: "Kal-El",
    img:
      "https://res.cloudinary.com/asalguerodev/image/upload/v1620275833/heros-app/dc-superman_zcd5ov.jpg",
  },
  {
    id: "dc-flash",
    superhero: "Flash",
    publisher: "DC Comics",
    alter_ego: "Jay Garrick",
    first_appearance: "Flash Comics #1",
    characters: "Jay Garrick, Barry Allen, Wally West, Bart Allen",
    img:
      "https://res.cloudinary.com/asalguerodev/image/upload/v1620275827/heros-app/dc-flash_lbqbwq.jpg",
  },
  {
    id: "dc-green",
    superhero: "Green Lantern",
    publisher: "DC Comics",
    alter_ego: "Alan Scott",
    first_appearance: "All-American Comics #16",
    characters:
      "Alan Scott, Hal Jordan, Guy Gardner, John Stewart, Kyle Raynor, Jade, Sinestro, Simon Baz",
    img:
      "https://res.cloudinary.com/asalguerodev/image/upload/v1620275829/heros-app/dc-green_epincq.jpg",
  },
  {
    id: "dc-arrow",
    superhero: "Green Arrow",
    publisher: "DC Comics",
    alter_ego: "Oliver Queen",
    first_appearance: "More Fun Comics #73",
    characters: "Oliver Queen",
    img:
      "https://res.cloudinary.com/asalguerodev/image/upload/v1620275826/heros-app/dc-arrow_hbg9ss.jpg",
  },
  {
    id: "dc-wonder",
    superhero: "Wonder Woman",
    publisher: "DC Comics",
    alter_ego: "Princess Diana",
    first_appearance: "All Star Comics #8",
    characters: "Princess Diana",
    img:
      "https://res.cloudinary.com/asalguerodev/image/upload/v1620275835/heros-app/dc-wonder_pxh6qv.jpg",
  },
  {
    id: "dc-martian",
    superhero: "Martian Manhunter",
    publisher: "DC Comics",
    alter_ego: "J'onn J'onzz",
    first_appearance: "Detective Comics #225",
    characters: "Martian Manhunter",
    img:
      "https://res.cloudinary.com/asalguerodev/image/upload/v1620275837/heros-app/dc-martian_lozvf4.jpg",
  },
  {
    id: "dc-robin",
    superhero: "Robin/Nightwing",
    publisher: "DC Comics",
    alter_ego: "Dick Grayson",
    first_appearance: "Detective Comics #38",
    characters: "Dick Grayson",
    img:
      "https://res.cloudinary.com/asalguerodev/image/upload/v1620275831/heros-app/dc-robin_ajaxc2.jpg",
  },
  {
    id: "dc-blue",
    superhero: "Blue Beetle",
    publisher: "DC Comics",
    alter_ego: "Dan Garret",
    first_appearance: "Mystery Men Comics #1",
    characters: "Dan Garret, Ted Kord, Jaime Reyes",
    img:
      "https://res.cloudinary.com/asalguerodev/image/upload/v1620275827/heros-app/dc-blue_gh9qfp.jpg",
  },
  {
    id: "dc-black",
    superhero: "Black Canary",
    publisher: "DC Comics",
    alter_ego: "Dinah Drake",
    first_appearance: "Flash Comics #86",
    characters: "Dinah Drake, Dinah Lance",
    img:
      "https://res.cloudinary.com/asalguerodev/image/upload/v1620275828/heros-app/dc-black_nvzvuv.jpg",
  },
  {
    id: "marvel-spider",
    superhero: "Spider Man",
    publisher: "Marvel Comics",
    alter_ego: "Peter Parker",
    first_appearance: "Amazing Fantasy #15",
    characters: "Peter Parker",
    img:
      "https://res.cloudinary.com/asalguerodev/image/upload/v1620275843/heros-app/marvel-spider_dm1bby.jpg",
  },
  {
    id: "marvel-captain",
    superhero: "Captain America",
    publisher: "Marvel Comics",
    alter_ego: "Steve Rogers",
    first_appearance: "Captain America Comics #1",
    characters: "Steve Rogers",
    img:
      "https://res.cloudinary.com/asalguerodev/image/upload/v1620275841/heros-app/marvel-captain_u8463f.jpg",
  },
  {
    id: "marvel-iron",
    superhero: "Iron Man",
    publisher: "Marvel Comics",
    alter_ego: "Tony Stark",
    first_appearance: "Tales of Suspense #39",
    characters: "Tony Stark",
    img:
      "https://res.cloudinary.com/asalguerodev/image/upload/v1620275838/heros-app/marvel-iron_qhpj40.jpg",
  },
  {
    id: "marvel-thor",
    superhero: "Thor",
    publisher: "Marvel Comics",
    alter_ego: "Thor Odinson",
    first_appearance: "Journey into Myster #83",
    characters: "Thor Odinson",
    img:
      "https://res.cloudinary.com/asalguerodev/image/upload/v1620275841/heros-app/marvel-thor_hmbbfk.jpg",
  },
  {
    id: "marvel-hulk",
    superhero: "Hulk",
    publisher: "Marvel Comics",
    alter_ego: "Bruce Banner",
    first_appearance: "The Incredible Hulk #1",
    characters: "Bruce Banner",
    img:
      "https://res.cloudinary.com/asalguerodev/image/upload/v1620275843/heros-app/marvel-hulk_wjjk9a.jpg",
  },
  {
    id: "marvel-wolverine",
    superhero: "Wolverine",
    publisher: "Marvel Comics",
    alter_ego: "James Howlett",
    first_appearance: "The Incredible Hulk #180",
    characters: "James Howlett",
    img:
      "https://res.cloudinary.com/asalguerodev/image/upload/v1620275842/heros-app/marvel-wolverine_hrvryb.jpg",
  },
  {
    id: "marvel-daredevil",
    superhero: "Daredevil",
    publisher: "Marvel Comics",
    alter_ego: "Matthew Michael Murdock",
    first_appearance: "Daredevil #1",
    characters: "Matthew Michael Murdock",
    img:
      "https://res.cloudinary.com/asalguerodev/image/upload/v1620275840/heros-app/marvel-daredevil_wcw1zz.jpg",
  },
  {
    id: "marvel-hawkeye",
    superhero: "Hawkeye",
    publisher: "Marvel Comics",
    alter_ego: "Clinton Francis Barton",
    first_appearance: "Tales of Suspense #57",
    characters: "Clinton Francis Barton",
    img:
      "https://res.cloudinary.com/asalguerodev/image/upload/v1620275836/heros-app/marvel-hawkeye_cmymws.jpg",
  },
  {
    id: "marvel-cyclops",
    superhero: "Cyclops",
    publisher: "Marvel Comics",
    alter_ego: "Scott Summers",
    first_appearance: "X-Men #1",
    characters: "Scott Summers",
    img:
      "https://res.cloudinary.com/asalguerodev/image/upload/v1620275835/heros-app/marvel-cyclops_suighl.jpg",
  },
  {
    id: "marvel-silver",
    superhero: "Silver Surfer",
    publisher: "Marvel Comics",
    alter_ego: "Norrin Radd",
    first_appearance: "The Fantastic Four #48",
    characters: "Norrin Radd",
    img:
      "https://res.cloudinary.com/asalguerodev/image/upload/v1620275840/heros-app/marvel-silver_nf3z1i.jpg",
  },
];
