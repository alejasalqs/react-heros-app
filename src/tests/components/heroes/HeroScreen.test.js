import React from "react";
import { shallow } from "enzyme";
import { HeroScreen } from "../../../components/heroes/HeroScreen";
import { MemoryRouter, Route } from "react-router";

describe("Pruebas en el componente <HeroScreen />", () => {
  const historyMock = {
    push: jest.fn(),
    location: {},
    listen: jest.fn(),
    createHref: jest.fn(),
    replace: jest.fn(),
    length: 10,
    goBack: jest.fn(),
  };

  // Esta prueba falla porque no se puede ver lo que hay dentro del wrapper
  test("Debe de mostrar el componente <Redirect /> si no hay parametros en el URL ", () => {
    // Se debe usar mount en este caso
    const wrapper = shallow(
      // InitialEntries es como el url que se va a enviar al ruter
      <MemoryRouter initialEntries={["/hero"]}>
        <HeroScreen history={historyMock} />
      </MemoryRouter>
    );
    expect(wrapper.find("Redirect").exists()).toBe(true);
  });

  test("Debe de  mostrar un heroe si el parametro existe y se encuentra", () => {
    // Se debe usar mount en este caso
    const wrapper = shallow(
      // InitialEntries es como el url que se va a enviar al ruter
      // usando el componente route podemos leer los paramentros
      // que se envian por url en el useParams()
      <MemoryRouter initialEntries={["/hero/marvel-spider"]}>
        <Route path="/hero/:heroId" component={HeroScreen} />
      </MemoryRouter>
    );

    expect(wrapper.find(".row").exists()).toBe(true);
  });

  test("Debe de regresar a la pantalla anterior con push ", () => {
    const historyMock = {
      push: jest.fn(),
      location: {},
      listen: jest.fn(),
      createHref: jest.fn(),
      replace: jest.fn(),
      length: 1,
      goBack: jest.fn(),
    };
    // Se debe usar mount en este caso
    const wrapper = shallow(
      // InitialEntries es como el url que se va a enviar al ruter
      // Aqui el componente es una funcion que devuelve el componente
      // A la funcion se le pasan los propr por paramentro
      <MemoryRouter initialEntries={["/hero/marvel-spider"]}>
        <Route
          path="/hero/:heroId"
          component={(props) => <HeroScreen history={historyMock} />}
        />
      </MemoryRouter>
    );
    wrapper.find("button").prop("onClick")();
    expect(historyMock.push).toHaveBeenCalled();
    expect(historyMock.goBack).not.toHaveBeenCalled();
  });

  test("Si el history es mayor a 2 no se debe llamar la funcion push ", () => {
    const historyMock = {
      push: jest.fn(),
      location: {},
      listen: jest.fn(),
      createHref: jest.fn(),
      replace: jest.fn(),
      length: 10,
      goBack: jest.fn(),
    };
    // Se debe usar mount en este caso
    const wrapper = shallow(
      // InitialEntries es como el url que se va a enviar al ruter
      // Aqui el componente es una funcion que devuelve el componente
      // A la funcion se le pasan los propr por paramentro
      <MemoryRouter initialEntries={["/hero/marvel-spider"]}>
        <Route
          path="/hero/:heroId"
          component={(props) => <HeroScreen history={historyMock} />}
        />
      </MemoryRouter>
    );
    wrapper.find("button").prop("onClick")();
    expect(historyMock.goBack).toHaveBeenCalled();
    expect(historyMock.push).not.toHaveBeenCalled();
  });

  test("Debe llamar el redirect si el heroe no existe ", () => {
    // Se debe usar mount en este caso
    const wrapper = shallow(
      // InitialEntries es como el url que se va a enviar al ruter
      // usando el componente route podemos leer los paramentros
      // que se envian por url en el useParams()
      <MemoryRouter initialEntries={["/hero/marvel-spider234234235344"]}>
        <Route path="/hero/:heroId" component={HeroScreen} />
      </MemoryRouter>
    );

    expect(wrapper.find("Redirect").exists()).toBe(true);
  });
});
