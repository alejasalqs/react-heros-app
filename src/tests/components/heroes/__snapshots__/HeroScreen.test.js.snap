// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <HeroScreen /> Debe de mostrar el componente <Redirect /> si no hay parametros en el URL  1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "canGo": [Function],
      "createHref": [Function],
      "entries": Array [
        Object {
          "hash": "",
          "key": "55benq",
          "pathname": "/hero",
          "search": "",
          "state": undefined,
        },
      ],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "index": 0,
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "55benq",
        "pathname": "/hero",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <HeroScreen
    history={
      Object {
        "createHref": [MockFunction],
        "goBack": [MockFunction],
        "length": 10,
        "listen": [MockFunction],
        "location": Object {},
        "push": [MockFunction],
        "replace": [MockFunction],
      }
    }
  />
</Router>
`;
